### ISSUE TITLE
[Design] Dynamic resource usage for GMP operator

### ISSUE THREAD

**Issue by pintohutch** [GoogleCloudPlatform/prometheus-engine#793]
Can we find ways to avoid OOM crashes in the gmp-operator? Maybe using a [VPA](https://gist.github.com/pintohutch/65bc578f1ca7f9d07ad44ff944168bb6)?

Acceptance criteria:
- Proposal with design and trade-offs

**Comment by clearclaw** [comment]
This weekend `gmp-operator` started crashlooping with, "no endpoints available for service gmp-operator".  While there were no OOMKill events, that turned out to be the cause.  Putting in a VPA addressed.

[Previous issue with rules-evaluator](https://github.com/GoogleCloudPlatform/prometheus-engine/issues/971)

- **ALL** deploys were broken due to gmp-operator's webhook being dead. 
- Really bad visibility on the OOMKill.  (I don't know why scheduler events weren't propagating, but they weren't)
- As with rules-evaluator: VPA FTW.

I see that the [rules-evaluator VPA](https://github.com/GoogleCloudPlatform/prometheus-engine/issues/975) hasn't moved either.  I can live lacking VPAs, but having the webhook fail closed for gmp-operator was excruciating.

**Comment by pintohutch** [comment]
Thanks for reporting @clearclaw - I apologize for the frustration.

With the operator being such a mission-critical binary in managed-collection, we can also explore failing open for our webhooks. Would that have helped in this situation over a VPA specifically?

**Comment by clearclaw** [comment]
Yes, in that deploys would have worked despite the failure.  Being unable
to eg  ship a key security fix due to GMP indigestion is not right.

The lack of built-in observability/alerting around GMP remains a concern
(gmp-operator or rules-evaluator previously here, but there's more
components).  If the observability stack fails, all the alarms should be
going off.  It would be really nice to see/have default Rules objects for
all of GMP.
-- JCL

**Comment by clearclaw** [comment]
Supplemental in case useful: 
![image](https://github.com/user-attachments/assets/9b123670-adb1-433c-955d-b7df6c317974)

**Comment by bwplotka** [comment]
+1, some way of handling this would be important, let's prioritize it.

VPA/limits is one thing, but perhaps additionally it would be to prioritize the code optimization phase, to check for low hanging fruits and start to queue things up (trading latency vs memory).

> Additionally, the lack of default observability/alerting around GMP
remains a concern (gmp-operator or rules-evaluator previously here,
but there's more components).  If the observability stack fails, all the
alarms should be going off.  It would be really nice to see/have
default Rules objects for all of GMP.  (Why are there no podmonitoring
objects for the GMP processes?)

For managed collection we don't do this, because we have our own alerting pipeline (however oriented on more fleet-wide situations and unknown situations). On top of that we historically didn't deploy those self-observability resources by default as everyone on GKE would need to pay for those extra metrics. We ship some example [configuration](https://github.com/GoogleCloudPlatform/prometheus-engine/blob/main/examples/self-pod-monitoring.yaml) anyone can deploy (on GKE std), though [missing operator](https://github.com/GoogleCloudPlatform/prometheus-engine/issues/1123).

To sum up, we have a bit to improve here, thanks for reporting & ideas. Perhaps it's time to add that `self-monitoring` feature/option to OperatorConfig but likely opt-in.

**Comment by clearclaw** [comment]
Making self-observation opt-in is sensible.  Something as simple as a doc
page or reference pointing at reference pod monitors and possible alert
rules under https://g.co/kgs/dDdszEW would help a lot.

In short, make it easy for people to do good things.
-- JCL
